### PR TITLE
Add a fast path for shaping ASCII text

### PR DESCRIPTION
**Pull request by mbrubeck**
On both my Linux laptop and iMac, this is about twice as fast as Harfbuzz text shaping on https://en.wikipedia.org/wiki/Barack_Obama.

I haven't tested this on any high-DPI (retina) displays, and I'm not 100% certain that the font unit scaling is correct there.

Depends on servo/core-text-rs#50.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11273)

### CODE REVIEW

File: components/gfx/platform/macos/font.rs [hunk 1]
```diff
@@ -155,12 +245,38 @@ impl FontHandleMethods for FontHandle {
...
+            let key = BigEndian::read_u32(&pairs[i * KERN_PAIR_LEN..]);
+            if key > query {
+                end = i;
+            } else if key < query {
+                start = i + 1;
+            } else {
+                let value = BigEndian::read_i16(&pairs[i * KERN_PAIR_LEN + 4..]);
+                trace!("Kerning for {}/{}: {:?}", first_glyph, second_glyph, value);
+                return value as f64 * table.scale;
+            }
```
Comment (pcwalton) [inline components/gfx/platform/macos/font.rs]: Can this use `Vec::binary_search()`?

Even if not, I still think I'd rather have this be a separate function.

File: components/gfx/text/shaping/harfbuzz.rs [hunk 2]
```diff
@@ -207,7 +201,45 @@ impl ShaperMethods for Shaper {
     /// Calculate the layout metrics associated with the given text when painted in a specific
     /// font.
     fn shape_text(&self, text: &str, options: &ShapingOptions, glyphs: &mut GlyphStore) {
+        let font = unsafe { &mut *self.font_and_shaping_options.font };
+        if options.script == Script::Latin &&
+            !options.flags.contains(RTL_FLAG) &&
+            font.handle.can_do_fast_kerning() &&
+            text.is_ascii()
+        {
```
Comment (pcwalton) [inline components/gfx/text/shaping/harfbuzz.rs]: Can you factor this out into a separate method?

File: components/gfx/platform/freetype/font.rs [hunk 3]
```diff
@@ -73,22 +75,16 @@ impl FontHandleMethods for FontHandle {
...
-                  handle: fctx.clone()
-              };
-              Ok(handle)
-            }
-            Err(()) => Err(())
-        };
+                  handle: fctx.clone(),
+                  can_do_fast_kerning: false,
+            };
+            handle.can_do_fast_kerning = handle.has_table(KERN) || !handle.has_table(GPOS);
```
Comment (pcwalton) [inline components/gfx/platform/freetype/font.rs]: Hmm. I would expect `handle.has_table(KERN) && !handle.has_table(GPOS)`. Am I missing something?

**Comment by mbrubeck** [comment]
Addressed review comments.

@pcwalton:

> Hmm. I would expect `handle.has_table(KERN) && !handle.has_table(GPOS)`. Am I missing something?

I assumed that if a font has both 'kern' and 'GPOS', then falling back to the 'kern' table would be sufficient for shaping ASCII-only strings.  It would be more conservative to use the slow path in this case, however.  This affects some commonly-used fonts like ArialMT and TimesNewRoman on Mac OS, and the DejaVu fonts on Linux.

Even if we implement partial GPOS support in the fast path, we'll need to make a similar assumption that ASCII characters aren't affected by any unsupported GPOS features (or, somehow scan the whole table to verify this).

We could also make this depend on the `text-rendering` CSS property…

**Comment by pcwalton** [comment]
@mbrubeck I was actually assuming we'd scan the table to verify this. I looked at the tables in FontForge and they seemed to pass the test from first glance.

But this can be a followup.

File: components/gfx/platform/freetype/font.rs (hunk 3 above)
Comment (mbrubeck) [inline components/gfx/platform/freetype/font.rs]: Switched to `has_table(KERN) && !has_table(GPOS)` for now.

**Comment by mbrubeck** [comment]
Changed to fall back to the slow path for now if the font has `GPOS` and/or `GSUB` tables.

We'll need to implement both `GPOS` pair positioning and `GSUB` ligation to get the fast path working for most Latin text.

**Comment by mbrubeck** [comment]
Whoops, messed up some tidy fixes. Will push a proper fix when I'm back at my computer.

File: components/gfx/font.rs [hunk 4]
```diff
@@ -24,6 +24,14 @@ use unicode_script::Script;
 use util::cache::HashCache;
 use webrender_traits;
 
+macro_rules! ot_tag {
+    ($t1:expr, $t2:expr, $t3:expr, $t4:expr) => (
+        (($t1 as u32) << 24) | (($t2 as u32) << 16) | (($t3 as u32) << 8) | ($t4 as u32)
+    );
+}
+
+pub const KERN: u32 = ot_tag!('k', 'e', 'r', 'n');
```
Comment (pcwalton) [inline components/gfx/font.rs]: Wish we had `const fn` :)

File: components/gfx/platform/freetype/font.rs [hunk 5]
```diff
@@ -74,11 +76,16 @@ impl FontHandleMethods for FontHandle {
...
 
         return create_face_from_buffer(ft_ctx, &template.bytes, pt_size).map(|face| {
-            let handle = FontHandle {
+            let mut handle = FontHandle {
                   face: face,
                   font_data: template.clone(),
                   handle: fctx.clone(),
+                  can_do_fast_shaping: false,
             };
+            // TODO: Implement basic support for GPOS and GSUB.
```
Comment (pcwalton) [inline components/gfx/platform/freetype/font.rs]: File an issue and link it here

Comment (mbrubeck) [inline components/gfx/platform/freetype/font.rs]: Done.

File: components/gfx/platform/macos/font.rs [hunk 6]
```diff
@@ -61,8 +64,105 @@ impl FontTableMethods for FontTable {
...
+        const FORMAT_0_HEADER_LEN: usize = 8;
+        {
+            let table = result.font_table.buffer();
+            let version = BigEndian::read_u16(table);
+            if version != 0 {
+                return None;
+            }
+            let num_subtables = BigEndian::read_u16(&table[2..]);
+            let mut start = 4;
+            for _ in 0..num_subtables {
```
Comment (pcwalton) [inline components/gfx/platform/macos/font.rs]: Kinda weird that the spec doesn't say what the subtable version number is supposed to be. Maybe add a TODO to maybe check this in the future?

**Comment by pcwalton** [comment]
Looks good, just had a couple of nits!

File: components/gfx/platform/macos/font.rs (hunk 6 above)
Comment (mbrubeck) [inline components/gfx/platform/macos/font.rs]: Done.

**Comment by mbrubeck** [comment]
@bors-servo r=pcwalton

* Fixed nits. Here we go…

**Comment by khaledhosny** [comment]
I think this is going backwards, Geko does not have a fast path and Blink is killing its fast path as well because it is proven to be a bad idea on the long run. HarfBuzz should be pretty fast when no OpenType layout is involved anyway, so if the code is slow then this is probably an issue in the Servo side.

CC @behdad.

**Comment by pcwalton** [comment]
The fast path in WebKit doesn't do kerning at all from what I can tell.

By design, our fast path only kicks in when we detect that there will be no quality difference between HarfBuzz and what we do. It's been shown to result in a 2x speedup, and I don't see anything wrong that we're doing in Servo that would cause a slowdown.

**Comment by khaledhosny** [comment]
I don’t know Rust that well, but the code seems to do the fast path when the font has a `kern` table and no `GPOS` or `GSUB`, but this is probably to match any AAT font, so on Mac OS X you won’t get any layout features for such fonts (not even kerning since will unlikely be a version 0 `kern` table). This would also match any Graphite font having a `kern` table (though I don’t know if it is common to have such a table in Graphite fonts).

Also if HarfBuzz path in Servo is slow, I think making it faster benefits everyone not only user of ASCII text and fonts with no OpenType layout table. Both Gecko and Blink cache the shaping result in various way to speed things up.

**Comment by mbrubeck** [comment]
This is still an experiment; it remains to be seen whether it will still be faster than HarfBuzz on typical fonts once we add GPOS and GSUB support. Perhaps it won't be. If we do find any significant wins, I'll definitely see if they they can be shared with HarfBuzz.

Part of the speedup here may be from better inlining. (We might also be able to get this from Harfbuzz by compiling it with Clang and using cross-language LTO, which is also something we plan to pursue.) Part of it is also from writing data directly into Servo's internal data structures without allocating any intermediate buffers.  And having a separate path for ASCII inside of Servo might be a win whether or not we it is using HarfBuzz.

> this is probably to match any AAT font, so on Mac OS X you won’t get any layout features for such fonts

Yeah, we ought to disable the fast path if any ASCII character is covered by a table that it doesn't support.  I'm working on this.

**Comment by behdad** [comment]
This is so not what we expect HarfBuzz clients to be doing...  My 0.02USD.

**Pull request merged**